pkg/reconciler/labeler/v2/accessors.go

```diff
@@ -57,9 +59,48 @@ func NewRevisionAccessor(
 	}
 }
 
-// get implements Accessor
-func (r *Revision) get(ns, name string) (kmeta.Accessor, error) {
-	return r.revisionLister.Revisions(ns).Get(name)
+// makeMetadataPatch makes a metadata map to be patched or nil if no changes are needed.
+func makeMetadataPatch(acc kmeta.Accessor, routeName *string) (map[string]interface{}, error) {
+	labels := map[string]interface{}{}
+
+	if err := addRouteLabel(acc, labels, routeName); err != nil {
+		return nil, err
+	}
+
+	if len(labels) > 0 {
+		meta := map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"labels": labels,
+			},
+		}
+		return meta, nil
+	}
+	return nil, nil
+}
+
+// addRouteLabel appends the route label to the list of labels if needed
+// or removes the label if routeName is nil.
+func addRouteLabel(acc kmeta.Accessor, diffLabels map[string]interface{}, routeName *string) error {
+	oldLabels := acc.GetLabels()
+	if routeName == nil { // remove the label
+		if len(oldLabels) != 0 && oldLabels[serving.RouteLabelKey] != "" {
+			diffLabels[serving.RouteLabelKey] = nil
+		}
+	} else { // add the label
+		if len(oldLabels) == 0 {
+			diffLabels[serving.RouteLabelKey] = routeName
+		} else {
+			// TODO(whaught): this restricts us to only one route -> revision
+			// We can move this to a comma separated list annotation and use the new routingState label.
+			if oldLabel := oldLabels[serving.RouteLabelKey]; oldLabel == "" {
+				diffLabels[serving.RouteLabelKey] = routeName
+			} else if oldLabel != *routeName {
+				return fmt.Errorf("resource already has route label %q, and cannot be referenced by %q", oldLabel, *routeName)
+			}
```

Comment:
How about:

```
if ol := ols[key]; ol == "" {
 // covers both empty map and key does not exist
  dl[key] = *routeName
} else if ol != *routeName {
   return error
}
```

Reply:
I think that panics if ols == nil which is why we had the len(old) == 0 check first

Reply:
No, map read operations are safe on nil maps.

Reply:
as you can see I'm terrified of panics. Cool let's ditch the empty checks